### PR TITLE
[RUNTIME] Broadcast based resharding

### PR DESCRIPTION
**Pull request by TarzanZhao**
Add feature: Broadcast based resharding. 

To enable this feature, we only need to set `global_config.resharding_mode="broadcast"`.

For more information of this method, please refer to [issues 416](https://github.com/alpa-projects/alpa/issues/416).

### CODE REVIEW

File: alpa/device_mesh.py [hunk 1]
```diff
@@ -498,6 +513,93 @@ def run_allgather_task(self, uuid, buffer_uuids):
...
+            elif global_rank == 0:
+                start_indices = tuple(o.start for o in tensor_slice)
+                to_use = jax_tensor_index(
+                    xla_buffer_to_jax_tensor(self.buffers[uuid]), start_indices,
+                    slice_shape)
+                to_use = jax_tensor_to_cupy(to_use)
+                slice_types.append("non_continuous_subset")
+            else:
+                to_use = device_put(
+                    jnp.ones(slice_shape, dtype=self.buffers[uuid].dtype),
```
Comment (ZYHowell) [inline alpa/device_mesh.py:542]: I think we've created a recv buffer by AllocZeroBufferExecutable

File: alpa/device_mesh.py [hunk 2]
```diff
@@ -498,6 +513,93 @@ def run_allgather_task(self, uuid, buffer_uuids):
...
+        for device_id, global_rank, tensor_slice in zip(devices_ids, devices_global_rank, tensor_slices):
+            uuid = uuids[device_id]
+            tensor_shape = self.buffers[uuid].shape
+            slice_shape = tuple(ind.stop - ind.start for ind in tensor_slice)
+            if is_continuous_subset(tensor_slice, tensor_shape):
+                # fast path, two cases: (1) same shape, (2) continuous subset.
+                if slice_shape != tensor_shape:
+                    ind, _ = infer_offset_and_n_elements(tensor_slice)
+                    slice_types.append("continuous_subset")
+                    to_use = (xla_buffer_to_cupy(self.buffers[uuid]), ind)
```
Comment (ZYHowell) [inline alpa/device_mesh.py:529]: two types of `to_use` make the code more complicated(more `if` like L549 and L560, `slice_types`). Is it possible to use `to_use`(buffer with offset, for `col.broadcast`) and `buffer`(buffer without offset, for `cupy_to_xla`).

**Review by ZYHowell — CHANGES_REQUESTED** [review]
LGTM. I'm not familiar with codes in `collective` so we need @zhisbug 's further review on this part

File: alpa/pipeline_parallel/stage_construction.py [hunk 3]
```diff
@@ -7,8 +7,8 @@
 
 import numpy as np
 import numba
-from ray.exceptions import RayActorError
 import tqdm
+from ray.exceptions import RayActorError
```
[the remark this answers is not in the document]
Comment (TarzanZhao) [reply alpa/pipeline_parallel/stage_construction.py:11]: Sry, what does r-s-t mean?

File: tests/test_broadcast_cross_mesh_resharding.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    strategy = CrossMeshCommunicator._generate_broadcast_resharding_strategy_by_loads(task_spec, src_loads, dst_loads)
+    task_spec.set_resharding_strategy(strategy)
+    # Resharding task. Compile send/recv from strategy and allgather.
+    collective_group = CollectiveGroup(task_spec.get_participant_device_strs(),
+                                       src_mesh, dst_mesh)
+    if global_config.eagerly_create_communicators:
+        collective_group.instantiate_now()
+    else:
+        collective_group.instantiate()
+    task = SymbolicBroadcastReshardingTask(task_spec, collective_group, src_mesh, dst_mesh)
```
[the remark this answers is not in the document]
Comment (TarzanZhao) [reply tests/test_broadcast_cross_mesh_resharding.py:54]: OK

File: alpa/device_mesh.py (hunk 1 above)
Comment (TarzanZhao) [reply alpa/device_mesh.py:542]: Yep, here I keep this functionality that eagerly puts buffer to devices. Do you mean I could just delete it in order to keep code simple because we will never use it.

File: alpa/device_mesh.py (hunk 2 above)
Comment (TarzanZhao) [reply alpa/device_mesh.py:529]: Sounds like a good idea, thx!

File: alpa/pipeline_parallel/stage_construction.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZYHowell) [reply alpa/pipeline_parallel/stage_construction.py:11]: not sure why you change the import order? it was in alphabetic

File: alpa/device_mesh.py (hunk 1 above)
Comment (ZYHowell) [reply alpa/device_mesh.py:542]: sry, I haven't noticed the likewise one in `recv_tile`. @zhisbug I remember we do not prefer buffer allocation with jax because it always compiles and is slow. Will this line have the problem? If not, just ignore this comment.

File: alpa/pipeline_parallel/cross_mesh_resharding.py [hunk 5]
```diff
@@ -700,6 +701,136 @@ def __str__(self):
...
+                self.communicator_params.add(param)
+        return self._broadcast_tasks
+
+    def _create_resharding_communicators(self):
+        """Create the NCCL communicators in advance."""
+        group_name = self.collective_group.group_name
+        for param in self.communicator_params:
+            task_dones = []
+            worker_to_devices_and_global_ranks = {}
+            world_size = len(param[1])
```
[the remark this answers is not in the document]
Comment (TarzanZhao) [reply alpa/pipeline_parallel/cross_mesh_resharding.py:801]: OK

File: alpa/pipeline_parallel/stage_construction.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TarzanZhao) [reply alpa/pipeline_parallel/stage_construction.py:11]: aha, the Linting feedback asks me to do that. No problem, let me change it back.

File: tests/test_cross_mesh_resharding.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 """Test cross-mesh resharding."""
+from email.policy import default
```
Comment (merrymercy) [inline tests/test_cross_mesh_resharding.py:2]: remove this

File: alpa/global_env.py [hunk 7]
```diff
@@ -108,6 +108,7 @@ def __init__(self):
         self.use_scatter_gather = True
         self.eagerly_create_communicators = True
         self.use_memzero_for_gradient_accumulation = False
+        self.resharding_mode = "default"  # or "broadcast"
```
Comment (merrymercy) [inline alpa/global_env.py:111]: Rename: `default`- > `send_recv`.
Later, we may use `broadcast` as the default

**Review by merrymercy — CHANGES_REQUESTED** [review]
How did you verify the performance?
Did you run these [instructions](https://github.com/alpa-projects/alpa/tree/main/benchmark/alpa#gpt-3)? How many GPUs do you test on?

File: alpa/collective/collective.py [hunk 8]
```diff
@@ -382,6 +382,43 @@ def broadcast(tensor, src_rank: int = 0, group_name: str = "default"):
...
+def broadcast_partialgpu(tensor_list,
+                         n_elements,
+                         comm_key,
+                         world_size,
+                         devices_ids,
+                         devices_global_rank,
+                         group_name: str = "default"):
+    """Broadcast the tensor from a source GPU to some other GPUs. 
+    This function is different from broadcast_multigpu that it only 
+    use a subset of gpus in one host.
```
[the remark this answers is not in the document]
Comment (TarzanZhao) [reply alpa/collective/collective.py:394]: OK

File: tests/test_cross_mesh_resharding.py (hunk 6 above)
Comment (TarzanZhao) [reply tests/test_cross_mesh_resharding.py:2]: OK

File: alpa/global_env.py (hunk 7 above)
Comment (TarzanZhao) [reply alpa/global_env.py:111]: OK

**Comment by TarzanZhao** [comment]
> 



> How did you verify the performance? Did you run these [instructions](https://github.com/alpa-projects/alpa/tree/main/benchmark/alpa#gpt-3)? How many GPUs do you test on?
> 
> For the CI problem, please rebase to the latest main branch.

For unittest,
image

For end-to-end, I use `python3 benchmark_3d.py --suite ` to test.

**Comment by zhisbug** [comment]
Tests are still failing?
Could you also post the performance numbers you got after you run the benchmark suites so we can take a look at? Numbers covering Both GPT and MoE are preferred

**Comment by TarzanZhao** [comment]
> Tests are still failing?
> Could you also post the performance numbers you got after you run the benchmark suites so we can take a look at? Numbers covering Both GPT and MoE are preferred

I do not know why `test_hlo_cost_model.py` fails here. It runs successfully in my experiment `python -m unittest test_hlo_cost_model.py`.

**Comment by zhisbug** [comment]
I just re-run it. If it still fails perhaps @merrymercy know the reason.

But you can try to get us some performance numbers as you did for UNet on MoE and GPT when broadcast is turned on

**Comment by TarzanZhao** [comment]
> 

The benchmark time depends on the number of iterations to run. My previous experiments are all done for 5 iterations. 

I am now doing a set of experiments and will upload the results later.

**Comment by zhisbug** [comment]
Sounds good. But note that The benchmark suite (benchmark_3d) will skip the first iteration (not using it for calculating the average time) as it is considered a warmup iter.

I think for broadcast, it should be the same with previous? The first iter is slow because of communicator construction, but the following items will be very stable?

**Comment by TarzanZhao** [comment]
I totally agree with you, based on my previous experiments.

**Comment by TarzanZhao** [comment]
With command `python3 benchmark_3d.py --suite gpt.perf_test_auto` on 16 gpus.

The first result is by broadcast, the second is by send_recv. 

image

**Comment by TarzanZhao** [comment]
For benchmarking moe, I think broadcast meets a deadlock. I am fixing it now. 

I will appreciate if you have ever met such error and could give some advices. 

image

**Comment by zhisbug** [comment]
> With command `python3 benchmark_3d.py --suite gpt.perf_test_auto` on 16 gpus.
> 
> The first result is by broadcast, the second is by send_recv. They have the same speed.
> 
> image

Cool, the results look great, thanks.

Additionally, for GPT-3, could you benchmark the following two cases:

32 GPUs: gpt     (1024, 1024, 5120, 48, 40, 51200)       (1, 4, 8)       (1, 4)  256     True    True    True    66.481s 0.096   15.364B 46.92   62.30   13.433G

64 GPUs: gpt Model Config: (1024, 1024, 8192, 48, 64) Parallel Config: (1, 4, 16) P-mesh shape: (1, 4) #Microbatch: 1024 Force Mapping: True Remat: True Reduce-scatter: True Mean Time: 87.268s Std Time: 0.288 #Params: 39.079B TFLOPs: 44.92 TFLOPs (ckpt): 59.74 Peak Mem: 13.609G overwrite_global_config_dict: None

And post the results here.

**Comment by TarzanZhao** [comment]
> > With command `python3 benchmark_3d.py --suite gpt.perf_test_auto` on 16 gpus.
> > The first result is by broadcast, the second is by send_recv. They have the same speed.
> > image
> 
> Cool, the results look great, thanks.
> 
> Additionally, for GPT-3, could you benchmark the following two cases:
> 
> 32 GPUs: gpt (1024, 1024, 5120, 48, 40, 51200) (1, 4, 8) (1, 4) 256 True True True 66.481s 0.096 15.364B 46.92 62.30 13.433G
> 
> 64 GPUs: gpt Model Config: (1024, 1024, 8192, 48, 64) Parallel Config: (1, 4, 16) P-mesh shape: (1, 4) #Microbatch: 1024 Force Mapping: True Remat: True Reduce-scatter: True Mean Time: 87.268s Std Time: 0.288 #Params: 39.079B TFLOPs: 44.92 TFLOPs (ckpt): 59.74 Peak Mem: 13.609G overwrite_global_config_dict: None
> 
> And post the results here.

OK, no problem.

File: alpa/device_mesh.py [hunk 9]
```diff
@@ -908,6 +1009,9 @@ def _launch_xla_servers(self):
                 # "NCCL_LAUNCH_MODE": "PARALLEL",
                 # "XLA_FLAGS": "--xla_dump_to=hlo --xla_dump_hlo_pass_re=.*"
                 # "NCCL_DEBUG": "INFO" if i == 0 else "VERSION",
+                # "NCCL_DEBUG_SUBSYS": "ALL",
+                "NCCL_ALGO": "Ring",
+                "NCCL_PROTO": "Simple",
```
Comment (merrymercy) [inline alpa/device_mesh.py:1014]: Is the CI error and NCCL deadlock due to this change?

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: These two env_vars are essential for broadcast to take advantage of nccl ring to really speed up.

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: And I further debug more deeply and find that the original error is `NCCL WARN Cuda failure 'out of memory'`.

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: image

Comment (merrymercy) [reply alpa/device_mesh.py:1014]: Try to decrease this value, so we leave more memory for other CUDA libraries (e.g., nccl, cublas, cudnn)
https://github.com/alpa-projects/alpa/blob/8a0753db1df49e51db15e823a6cf72d6a32f927b/alpa/global_env.py#L64-L65

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: Thanks!

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: Really appreciate!!! That works!

Comment (zhisbug) [reply alpa/device_mesh.py:1014]: it means broadcast algorithms require more memory?

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: I guess it means broadcast needs larger buffer?

Comment (zhisbug) [reply alpa/device_mesh.py:1014]: #432 can solve this memory issue...

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: Cool, Thanks!

Comment (TarzanZhao) [reply alpa/device_mesh.py:1014]: After changing from 0.9 to 0.87 for `self.xla_client_mem_fraction`, the above error disappears. 

The following are results for `python3 benchmark_3d.py --suite moe.perf_test_auto`. The first is by broadcast, the second is by send_recv. 

image

Broadcast uses less memory. I believe it is because tensor sent by broadcast is more complete, thus more likely to be contiguous, thus requires less extra memcpy in https://github.com/alpa-projects/alpa/blob/3416c2c0ced6f222425867952f355fd514ac1207/alpa/device_mesh.py#L332-L338

**Comment by TarzanZhao** [comment]
> > With command `python3 benchmark_3d.py --suite gpt.perf_test_auto` on 16 gpus.
> > The first result is by broadcast, the second is by send_recv. They have the same speed.
> > image
> 
> Cool, the results look great, thanks.
> 
> Additionally, for GPT-3, could you benchmark the following two cases:
> 
> 32 GPUs: gpt (1024, 1024, 5120, 48, 40, 51200) (1, 4, 8) (1, 4) 256 True True True 66.481s 0.096 15.364B 46.92 62.30 13.433G
> 
> 64 GPUs: gpt Model Config: (1024, 1024, 8192, 48, 64) Parallel Config: (1, 4, 16) P-mesh shape: (1, 4) #Microbatch: 1024 Force Mapping: True Remat: True Reduce-scatter: True Mean Time: 87.268s Std Time: 0.288 #Params: 39.079B TFLOPs: 44.92 TFLOPs (ckpt): 59.74 Peak Mem: 13.609G overwrite_global_config_dict: None
> 
> And post the results here.

For broadcast experiment result on 32 gpu, broadcast is little bit faster. 
image

For broadcast experiment result on 64 gpu, broadcast is little bit faster and consume less memory. 

image

**Review by ZYHowell — APPROVED** [review]

File: tests/test_cross_mesh_resharding.py [hunk 10]
```diff
@@ -242,14 +276,32 @@ def test_8gpu_2_dim_allgather(self):
             [NoSharding(), NoSharding(),
              NoSharding()], [Replicated(4)])
         self.run_resharding_task(src_shape, dst_shape, src_spec, dst_spec,
-                                  tensor_shape)
+                                  tensor_shape, resharding_mode="send_recv")
+        self.run_resharding_task(src_shape, dst_shape, src_spec, dst_spec,
+                                  tensor_shape, resharding_mode="broadcast")
+
+    def test_4gpu_broadcast(self):
```
Comment (merrymercy) [inline tests/test_cross_mesh_resharding.py:283]: It is 8 gpu

File: tests/test_cross_mesh_resharding.py [hunk 11]
```diff
@@ -242,14 +276,32 @@ def test_8gpu_2_dim_allgather(self):
...
+                                  tensor_shape, resharding_mode="broadcast")
 
 def suite():
     suite = unittest.TestSuite()
-    suite.addTest(ReshardingTest("test_4gpu_send_recv"))
+    suite.addTest(ReshardingTest("test_4gpu_all"))
     suite.addTest(ReshardingTest("test_4gpu_allgather"))
-    suite.addTest(ReshardingTest("test_8gpu_2_dim_allgather"))
+    suite.addTest(ReshardingTest("test_8gpu_2_dim_allgather_broadcast"))
+    suite.addTest(ReshardingTest("test_4gpu_broadcast"))
```
Comment (merrymercy) [inline tests/test_cross_mesh_resharding.py:304]: The names and organizations of these unit tests are very confusing. I suggest the following modifications

1. keep all old test cases unchanged and use `send_recv` in the old test cases
2. add two new cases `test_4gpu_broadcast` and  `test_8gpu_broadcast`

File: alpa/device_mesh.py [hunk 12]
```diff
@@ -291,6 +299,13 @@ def init_collective_group(world_size, rank, backend, group_name):
                                   backend=backend,
                                   group_name=group_name)
 
+    @staticmethod
+    def init_broadcast_communicator(group_name, comm_key, world_size, device_ids, devices_global_rank, nccl_uid):
```
Comment (merrymercy) [inline alpa/device_mesh.py:303]: Does this solve the previous to-do item (i.e., creating communicators in advance)?

**Review by merrymercy — CHANGES_REQUESTED** [review]
The perf looks good to me. Since this PR provides a new mode instead of replacing the default mode, we can try to merge this. Please fix the remaining issues.

For the CI problem, I don't know why. Could you try
```
python3 run_all.py --filter test_hlo_cost_model.py
```
```

File: alpa/device_mesh.py [hunk 13]
```diff
@@ -908,6 +1009,9 @@ def _launch_xla_servers(self):
                 # "NCCL_LAUNCH_MODE": "PARALLEL",
                 # "XLA_FLAGS": "--xla_dump_to=hlo --xla_dump_hlo_pass_re=.*"
                 # "NCCL_DEBUG": "INFO" if i == 0 else "VERSION",
+                # "NCCL_DEBUG_SUBSYS": "ALL",
+                "NCCL_ALGO": "Ring",
```
Comment (merrymercy) [inline alpa/device_mesh.py:1013]: Revert these changes. If they are required for broadcast, only set the when `global_config.resharding_mode == "broadcast"`

**Comment by TarzanZhao** [comment]
> The perf looks good to me. Since this PR provides a new mode instead of replacing the default mode, we can try to merge this. Please fix the remaining issues.
> 
> For the CI problem, I don't know why. Could you try
> 
> ```
> python3 run_all.py --filter test_hlo_cost_model.py
> ```

This command runs successfully for both two resharding modes in the following screenshot on one p3.16xlarge instance. 
image

File: alpa/device_mesh.py (hunk 12 above)
Comment (TarzanZhao) [reply alpa/device_mesh.py:303]: Yes, this part of code solves the to-do item,that indeed creates communicators in advance.

File: tests/test_cross_mesh_resharding.py (hunk 10 above)
Comment (TarzanZhao) [reply tests/test_cross_mesh_resharding.py:276]: OK

File: tests/test_cross_mesh_resharding.py (hunk 11 above)
Comment (TarzanZhao) [reply tests/test_cross_mesh_resharding.py:318]: OK

File: alpa/device_mesh.py (hunk 13 above)
Comment (TarzanZhao) [reply alpa/device_mesh.py]: OK

**Comment by TarzanZhao** [comment]
@merrymercy I have finshed the requested changes from you. For CI problem, It runs successfully in my experiments mentioned above, I am still confused now.

**Review by merrymercy — APPROVED** [review]

**Comment by merrymercy** [comment]
@TarzanZhao  This is merged! I will also try to run more tests and benchmarks.

**Comment by TarzanZhao** [comment]
Thanks!

**Comment by zhisbug** [comment]
Side note: @TarzanZhao on MoE, this broadcast feature will slow down (~2TFlops) on 16 GPUs, and match send/recv on 8  GPUs

**Comment by TarzanZhao** [comment]
I have uploaded broadcast result on moe here before. But I do not know why it does not exist now.

**Comment by TarzanZhao** [comment]
Some experiments results TFLOP(ckpt) for broadcast.

|     | 8 GPU  | 16 GPU | 32 GPU | 64 GPU |
|  ----  | ----  | ---- | ----  | ----  |
|  gpt | | 61.54 | 63.36 | 62.05 |
| moe  | 48.78 | 48.56 | | |
